### PR TITLE
DEPRECATE `private_token` in `.gitreview`; use git credential helper

### PR DESCRIPTION
**Pull request by thcipriani**
Current status: `private_token` is stored in `.gitreview` or in git config files as plaintext.

After this patch: storing a `private_token` in `.gitreview` is DEPRECATED and will print a warning.

And users will be able to store their `private_token` where it's only accessible using their git-credential helper of choice.

Change-Id: Ib1c30fa80f143c00e06829a64aa5bf69c121c018

### CODE REVIEW

**Comment by thcipriani** [comment]
hrm, unsure how to re-run tests following the `--force-with-lease` update 🤔

File: gerritlab/global_vars.py
Comment (yaoyuannnn) [inline gerritlab/global_vars.py]: Can we throw an error instead if this is to be deprecated?

File: gerritlab/global_vars.py
Comment (yaoyuannnn) [inline gerritlab/global_vars.py]: Use a separate variable instead of specializing a key in `self._git_credentials`.

File: gerritlab/global_vars.py
Comment (yaoyuannnn) [inline gerritlab/global_vars.py]: `remote` (instead of `gitreview_remote`)?

**Comment by yaoyuannnn** [comment]
> hrm, unsure how to re-run tests following the `--force-with-lease` update 🤔

I assume the CI will be triggered after a force push?

**Review by yaoyuannnn — COMMENTED** [review]
Thanks for doing this! Left a few minor comments.

File: README.md
Comment (dancysoft) [inline README.md]:
```suggestion
* `host`: The base URL of the GitLab server.
```

File: README.md
Comment (dancysoft) [inline README.md]:
```suggestion
A more secure option is to use a **git credential store**. This option enables you
```

File: README.md
Comment (dancysoft) [inline README.md]: Maybe use a generic example username.

File: README.md
Comment (dancysoft) [inline README.md]: Link to https://git-scm.com/docs/git-credential-store

Clarify that using the default store is **not** more secure than storing the token in git config.  

This section would be more useful if it showed a truly secure approach but I understand that there are a lot of options for doing that.

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]:
```suggestion
    def __init__(self, host: str) -> None:
```

File: gerritlab/global_vars.py
Comment (dancysoft) [reply gerritlab/global_vars.py]: Agreed.  Also, from a code-reading perspective it would be nice if the check for the fetch flag was in the same function that sets the flag.

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]: I don't have a credentials store configured so when I run `git lab` with this code in place, it ends up prompting me for username/password where it didn't before (because I have the token in my ~/.gitconfig).  Because of this I think checking the credential store should be the last thing that is attempted.

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]: This captures (and makes invisible) any interactive prompts the credential helper might be writing.

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]: Please trap `KeyboardInterrupt` and return "" when it is raised.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Done! Hard deprecated now.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Done.

File: README.md
Comment (thcipriani) [reply README.md]: Rethought this section a bit.

File: README.md
Comment (thcipriani) [reply README.md]: I clarified in the latest version that the git-credential-store should be avoided as a credential helper.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Moved to the last option.

Comment (thcipriani) [reply gerritlab/global_vars.py]: Done.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: This is true, though it seems most credential helpers are meant to be used in this way (i.e., input from stdin, username/password on stdout).

The default git prompt uses the tty directly. Or it uses a graphical askpass program.

Open to better ways of resolving this. But this works fine with all mechanisms I tested:
- pass-git-helper
- git-credential-manager (using libsecret)
- cache
- store

File: README.md
Comment (dancysoft) [inline README.md:75]: This section looks great!

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]: Consider using `text=True` instead of `.encode()` (easier to read and consistent with the code in `_call_git_credential_fill`)

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py]: We shouldn't approve until we know that the credentials really worked.

File: gerritlab/global_vars.py
Comment (dancysoft) [reply gerritlab/global_vars.py]: Agreed that the helper should be using the tty for prompting.  I'll look deeper into the credential helper that I'm using (https://github.com/develerik/git-credential-1password) to see if it's doing something wrong.

Comment (dancysoft) [reply gerritlab/global_vars.py]: Summary:  `capture_output=True` is capturing both stdout (ok) and stderr (not okay.. that's where prompts and other information output is written).   Use `stdout=subprocess.PIPE` instead to leave stderr unaffected.

File: README.md
Comment (dancysoft) [inline README.md]: Instead of including the protocol details, I suggest just describing how the credentials must be named/identified/titled in your password, with an example.

Comment (thcipriani) [reply README.md]: Good call, attempted that in latest revision.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Tried an implementation of that in the latest revision, thoughts?

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Done!

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Doh! Thanks for this <3 Fixed now.

**Comment by thcipriani** [comment]
Noteworthy in latest revision: I got a chance to test with the macos keychain today (`git credential.helper osxkeychain`)—it refused to save or use credentials using my made up protocol `gerritlab://` AND my url using `gerritlab@<host>`.

I've opted to use `https://<host>/_gerritlab` as the path for the token in this revision—this makes `osxkeychain` happy, AFAICT. This should still enable users to store alternative tokens for the same host in the same keychain WITHOUT conflicts for gerritlab.

**Comment by supertassu** [comment]
Does this work with the credential helper [`glab`](https://gitlab.com/gitlab-org/cli) installs?

File: README.md
Comment (dancysoft) [inline README.md]: The `host` entry in `.gitreview` is a URL, so I would say that it expects  `<host>` + `/_gerritlab`

File: README.md
Comment (dancysoft) [inline README.md]: host will be something like `https://gitlab.com`

File: gerritlab/git_credentials.py
Comment (dancysoft) [inline gerritlab/git_credentials.py:1]: This is what happens to me:

```
dancy@base:~/src/wmf/hacking/deleteme$ git lab
[ERROR] 2023/09/11 08:24:05 "https://gitlab.wikimedia.org/_gerritlab" isn't an item in the "Private" vault. Specify the item with its UUID, name, or domain.
Username for 'https://https://gitlab.wikimedia.org/_gerritlab':
```
The ERROR is from the 1password credential helper.  The username prompt is from `git credentials fill`.  Note the extra `https://`.

File: gerritlab/global_vars.py
Comment (dancysoft) [reply gerritlab/global_vars.py]: Looks reasonable.

File: README.md
Comment (thcipriani) [reply README.md]: Fixed!

File: README.md
Comment (thcipriani) [reply README.md]: I also removed the `_gerritlab` thing. Different credential helpers get confused by different things. Just using https + host now. Updated docs to reflect that.

**Comment by thcipriani** [comment]
> Does this work with the credential helper [`glab`](https://gitlab.com/gitlab-org/cli) installs?

Yep, latest patchset worked with `git config credential.helper '!glab auth git-credential "$@"'`

File: gerritlab/git_credentials.py
Comment (dancysoft) [inline gerritlab/git_credentials.py]: Maybe: parse the host URL and use the protocol from it instead of hard-coding `https`.

File: gerritlab/global_vars.py
Comment (dancysoft) [inline gerritlab/global_vars.py:96]:
```suggestion
    raise SystemExit(f"Unable to find private token for {host_netloc}")
```

**Review by dancysoft — COMMENTED** [review]
This version works for me with my modified 1Password credential helper.  If I don't already have a matching entry in a vault, `git credential fill` prompts for username and password and the subsequent `git credential approve` saves the creds as a permanent entry in the vault.

File: gerritlab/git_credentials.py
Comment (thcipriani) [reply gerritlab/git_credentials.py]: Done.

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py:96]: Changed in latest.

**Comment by dancysoft** [comment]
LGTM! @yaoyuannnn do you want to take a final look before merge happens?

File: gerritlab/global_vars.py
Comment (yaoyuannnn) [inline gerritlab/global_vars.py]: nit: 3->4

**Review by yaoyuannnn — APPROVED** [review]
LGTM! Thanks guys for adding this!

File: gerritlab/global_vars.py
Comment (thcipriani) [reply gerritlab/global_vars.py]: Good catch! Done.

**Comment by dancysoft** [comment]
Let's do this thing!

**Pull request merged**
